### PR TITLE
fix(template): process manifests in file path order, then in order found in each file (before sorting manifests)

### PR DESCRIPTION
**Pull request by AndiDog**
Previously, `helm template` would go through e.g. `templates/*.yml` in unspecified order, and the same issue applied when looping through the manifests within each file. That's because Go does not guarantee ordering of a map ranges, but Helm loops over maps. The unfortunate result: each call to `helm template` might produce a different order of manifest objects. While this surely went unnoticed for so long because most people use Helm as deployment operator, it's pretty bad for people who either commit the `helm template` output somewhere, or even auto-commit in a CI/CD pipeline – producing random results would make the diffs unreadable because objects keep moving to somewhere in the generated output.

This PR adds a fix and test (+ 100 test repeats because with the buggy implementation, the test could have randomly produced the expected result if run only once).

### CODE REVIEW

**Comment by bacongobbler** [comment]
is there a bug in teh queue this PR should reference?

**Comment by AndiDog** [comment]
I created #6843 for completeness' sake 😉

**Comment by thomastaylor312** [comment]
See https://github.com/helm/helm/issues/6843#issuecomment-548143055 for additional discussion, but we are labeling this as a feature and adding to the 3.1 milestone

**Comment by AndiDog** [comment]
Renamed the commit to call it a "feature", not a "fix"

File: pkg/releaseutil/manifest.go
Comment (bacongobbler) [inline pkg/releaseutil/manifest.go]: This would be considered a breaking change as the expected output has been modified. The zero-index used to be `manifest-0`, but now it'll be `manifest-0000000000`. Users may have calling code expecting the former.

Additionally, that means this function has been limited to sort a maximum of 10 billion manifests, whereas before it could sort `math.MaxInt`. Not a big deal as I don't think anyone was sorting that amount of manifests, but regardless it's still a behavioural change.

A non-breaking change would be to replace `sort.Strings` with `sort.Sort` in `SortManifests`, and implement a sorter for the output of `SplitManifests`. That way, `sort.Sort` sorts the manifests in the expected order during SortManifests.

**Review by bacongobbler — CHANGES_REQUESTED** [review]

File: cmd/helm/template_test.go
Comment (AndiDog) [reply cmd/helm/template_test.go]: Before the fix, Helm took files in random map order and could therefore produce indeterministic results and even give the expected, sorted result by chance. I've reduced the repeats to 10 and explained the rationale. Once flaky, always flaky, so not much we can do here if we want to ensure previous behavior doesn't happen anymore.

File: pkg/releaseutil/manifest.go
Comment (AndiDog) [reply pkg/releaseutil/manifest.go]: Applied as suggested, sorting by the part after `manifest-`.

In addition, I changed the kind sorter to not sort by name (if other attributes are equal), since that would be against the idea of keeping the original file order as suggested in the PR.

**Review by bacongobbler — APPROVED** [review]
LGTM! Thanks for working on this. Needs another maintainer's review before we can merge this.

**Review by thomastaylor312 — APPROVED** [review]
This is looking great! Thank you for the well written test coverage

**Comment by tmshort** [comment]
Will this be in 3.0.3?

**Comment by hickeyma** [comment]
@tmshort As it is a feature, it will probably be in 3.1.0

**Comment by tmshort** [comment]
I would argue that it's a bug; without this fix ("fix" is in the title), it prevents reasonably diff'ing the output of different runs of `helm template`. I'm seen cases where identical input to `helm template` (specifically for istio) have differences of several thousand lines due to the mis-ordering.

**Comment by AndiDog** [comment]
I'm not seeing consistent output in 3.1.0-rc.3.

@bacongobbler @thomastaylor312 Was this feature actually merged into the release candidate?

**Comment by AndiDog** [comment]
Seems to work in v3.1.0

**Comment by hickeyma** [comment]
@AndiDog It looks like it was added as per the change log: https://github.com/helm/helm/releases/tag/v3.1.0

**Pull request merged**
